### PR TITLE
Aruha 535 cleanup timelines

### PR DESCRIPTION
**Pull request by v-stepanov**
https://techjira.zalando.net/browse/ARUHA-535

### CODE REVIEW

File: database/nakadi/10_data/04_tables/05_timeline.sql [hunk 1]
```diff
@@ -8,6 +8,7 @@ CREATE TABLE zn_data.timeline (
   tl_switched_at     TIMESTAMP WITH TIME ZONE DEFAULT NULL,
   tl_cleanup_at      TIMESTAMP WITH TIME ZONE DEFAULT NULL,
   tl_latest_position JSONB                    DEFAULT NULL,
+  tl_deleted         BOOLEAN                  DEFAULT FALSE,
```
Comment (v-stepanov) [inline database/nakadi/10_data/04_tables/05_timeline.sql]: Well, right now when event-type is deleted - it is removed from DB, so it will make no sense to keep timelines for ET that was deleted. 
But when ET still exists I think it's better to keep all timelines for the history (at least for now when timeline feature is not really tested well)

File: src/acceptance-test/java/org/zalando/nakadi/repository/db/TimelineDbRepositoryTest.java [hunk 2]
```diff
@@ -116,20 +88,46 @@ public void testListTimelinesOrdered()
...
+
+        insertTimeline(tomorrow.toDate(), true, 0);
+        insertTimeline(tomorrow.toDate(), false, 1);
+        insertTimeline(yesterday.toDate(), true, 2);
+        final Timeline t4 = insertTimeline(yesterday.toDate(), false, 3);
+        final Timeline t5 = insertTimeline(twoDaysAgo.toDate(), false, 4);
+
+        final List<Timeline> expiredTimelines = tRepository.getExpiredTimelines();
+
+        Assert.assertEquals(2, expiredTimelines.size());
```
Comment (v-stepanov) [inline src/acceptance-test/java/org/zalando/nakadi/repository/db/TimelineDbRepositoryTest.java]: Good point 👍 
I now added explicit check for NULL and extended test https://github.com/zalando/nakadi/pull/611/commits/b0f8b6b49c16ad5b4f6a33e9599757faaa6d0172

File: src/acceptance-test/java/org/zalando/nakadi/repository/db/TimelineDbRepositoryTest.java [hunk 3]
```diff
@@ -116,20 +88,46 @@ public void testListTimelinesOrdered()
...
+        Assert.assertEquals(ImmutableList.of(t4, t5), expiredTimelines);
+    }
+
+    private Timeline insertTimelineWithOrder(final int order) {
+        return tRepository.createTimeline(createTimeline(
+                storage, UUID.randomUUID(), order, "test_topic", testEt.getName(),
+                new Date(), null, null, null));
+    }
+
+    private Timeline insertTimeline(final Date cleanupDate, final boolean deleted, final int order) {
```
Comment (v-stepanov) [inline src/acceptance-test/java/org/zalando/nakadi/repository/db/TimelineDbRepositoryTest.java]: I could not find a good name for this method, so I renamed the other one :)

File: src/acceptance-test/java/org/zalando/nakadi/service/job/ExclusiveJobWrapperAT.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+
+    @Test
+    public void whenExecuteJobThenOk() throws Exception {
+        jobWrapper.runJobLocked(dummyJob);
+        assertThat(jobExecuted.get(), is(true));
+
+        assertThat("lock node should be removed", CURATOR.checkExists().forPath(lockPath), is(nullValue()));
+
+        final byte[] data = CURATOR.getData().forPath(latestPath);
+        final DateTime lastCleaned = objectMapper.readValue(new String(data, Charsets.UTF_8), DateTime.class);
```
Comment (v-stepanov) [inline src/acceptance-test/java/org/zalando/nakadi/service/job/ExclusiveJobWrapperAT.java]: I think the current "standard" is to use guava Charsets class.
Changed https://github.com/zalando/nakadi/pull/611/commits/c869f991250d95cd892d4e79300f3ca724b9466d

File: src/acceptance-test/java/org/zalando/nakadi/webservice/TimelinesControllerAT.java [hunk 5]
```diff
@@ -6,28 +6,36 @@
...
+    private static final long RETENTION_TIME_MS = Duration.of(2, DAYS).toMillis();
 
-    @BeforeClass
-    public static void setUp() throws JsonProcessingException {
-        NakadiTestUtils.createEventTypeInNakadi(eventType);
-    }
+    private final EventType eventType;
 
-    @AfterClass
-    public static void tearDown() {
```
Comment (v-stepanov) [inline src/acceptance-test/java/org/zalando/nakadi/webservice/TimelinesControllerAT.java]: I don't really see a reason to cleanup. In out ATs we usually don't cleanup because it will just add extra time to execution and it is not really needed as these tests are never run on real DBs. 
The only exception is UserJourneyAT, we run it on real environments and there we do cleanup after test.

File: src/main/java/org/zalando/nakadi/service/job/ExclusiveJobWrapper.java [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+
+        if (!timeToRunAction()) {
+            return;
+        }
+        if (!createJobLockNode()) {
+            return;
+        }
+        // we need to check it again after we acquired the lock
+        if (!timeToRunAction()) {
+            return;
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/job/ExclusiveJobWrapper.java]: Holy crap! Good that you noticed that 👍 
I moved this now under the `try` so that lock is released by `finally` block. 
https://github.com/zalando/nakadi/pull/611/commits/abbee68f6446b6e7484852109861eca885c0331f

File: src/main/java/org/zalando/nakadi/service/job/ExclusiveJobWrapper.java [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+        }
+        try {
+            log.info("Execution of job will be performed on this instance");
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/job/ExclusiveJobWrapper.java]: Fixed https://github.com/zalando/nakadi/pull/611/commits/ac02cc6e1304e32d75a080522a1ad581a2bc5096

File: database/nakadi/10_data/04_tables/05_timeline.sql (hunk 1 above)
Comment (lmontrieux) [inline database/nakadi/10_data/04_tables/05_timeline.sql]: That's a good idea, but unfortunately the column `st_id` in table `timeline` is a foreign key to `st_id` in table `storage`. So we wouldn't be able to delete a storage, if there is one deleted timeline left on the storage.

Comment (v-stepanov) [inline database/nakadi/10_data/04_tables/05_timeline.sql]: Ok, this is an interesting point. But I think that it should be fixed from other side. When the  storage is deleted it should also delete all timelines of this storage. And it should also check if at least one timeline is still used - then it should forbid deletion of storage.
I will add these changes to storage deletion.

**Comment by lmontrieux** [comment]
👍

**Comment by v-stepanov** [comment]
@lmontrieux I reworked deletion of storage. So now when storage is deleted - it also removes all timelines with deleted=true.

**Comment by lmontrieux** [comment]
@v-stepanov would you like to add an acceptance test that deletes a storage containing exclusively deleted timelines?

**Comment by lmontrieux** [comment]
👎  I'm blocking this for now, after this morning's discussion with @antban during SU.

We need to discuss what the best strategy is wrt deletion of timelines, storages, and event types. If we decide to mark timelines as deleted instead of actually deleting them, then perhaps it makes sense to do the same with storages and/or event types.

**Comment by lmontrieux** [comment]
👎

**Comment by lmontrieux** [comment]
@v-stepanov So, we discussed this on Friday, and decided that we will only allow a storage to be deleted if there are no timelines. Even if all timelines in the storage are marked as deleted, we do not  delete the storage.

File: src/acceptance-test/java/org/zalando/nakadi/webservice/TimelinesControllerAT.java (hunk 5 above)
Comment (adyach) [inline src/acceptance-test/java/org/zalando/nakadi/webservice/TimelinesControllerAT.java]: I don not agree with this, If test initialises resource it has to free them. It will avoid situations when tests work with the event type, which was modified in other test.

File: src/main/java/org/zalando/nakadi/service/StorageService.java [hunk 8]
```diff
@@ -87,24 +105,33 @@ public StorageService(final ObjectMapper objectMapper,
...
-            if (storageDbRepository.isStorageUsed(id)) {
-                return Result.forbidden("Storage " + id + " is in use");
-            }
-            storageDbRepository.deleteStorage(id);
-        } catch (InternalNakadiException e) {
+            transactionTemplate.execute(action -> {
+                storageDbRepository.deleteStorage(id);
+                return null;
+            });
+        } catch (final NoStorageException e) {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/StorageService.java]: Do you think we can continue with the exceptions refactoring you started in repository level and throw these exceptions to controller level catching them with exception handling mechanism?

File: src/main/java/org/zalando/nakadi/service/StorageService.java [hunk 9]
```diff
@@ -87,24 +105,33 @@ public StorageService(final ObjectMapper objectMapper,
...
         try {
-            if (!storageDbRepository.getStorage(id).isPresent()) {
-                return Result.notFound("No storage with ID " + id);
-            }
-            if (storageDbRepository.isStorageUsed(id)) {
-                return Result.forbidden("Storage " + id + " is in use");
-            }
-            storageDbRepository.deleteStorage(id);
-        } catch (InternalNakadiException e) {
+            transactionTemplate.execute(action -> {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/StorageService.java]: It is not clear why you do one operation in transaction. Could you elaborate on that?

File: src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java [hunk 10]
```diff
@@ -184,6 +188,7 @@ private void switchTimelines(final Timeline activeTimeline, final Timeline nextT
                     final Timeline.StoragePosition storagePosition =
                             topicRepositoryHolder.createStoragePosition(activeTimeline);
                     activeTimeline.setLatestPosition(storagePosition);
+                    scheduleTimelineCleanup(activeTimeline);
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java]: Shouldn't it be done when we create topic?

File: src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+
+    @Scheduled(
+            fixedDelayString = "${nakadi.jobs.checkRunMs}",
+            initialDelayString = "${random.int(${nakadi.jobs.checkRunMs})}")
+    public void cleanupTimelines() {
+        try {
+            jobWrapper.runJobLocked(() ->
+                    timelineDbRepository.getExpiredTimelines().stream()
+                            .forEach(timeline -> {
+                                deleteTimelineTopic(timeline);
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java]: In my opinion we should stop cleaning up if deletion of the topic is failed and also if we deleted the topic, but the timeline deletion is failed, what then?

File: src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java [hunk 12]
```diff
@@ -197,6 +202,20 @@ private void switchTimelines(final Timeline activeTimeline, final Timeline nextT
...
     }
 
+    private void scheduleTimelineCleanup(final Timeline timeline) throws InconsistentStateException {
+        try {
+            final EventType eventType = eventTypeCache.getEventType(timeline.getEventType());
+            final Long retentionTime = eventType.getOptions().getRetentionTime();
+            if (retentionTime == null) {
+                throw new InconsistentStateException("Event type should has information about its retention time");
+            }
+            final Date cleanupDate = new Date(System.currentTimeMillis() + retentionTime);
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java]: There is a possibility to change retention time of the event type. CleanupUpAt has to be changed accordingly.

File: src/main/java/org/zalando/nakadi/service/job/ExclusiveJobWrapper.java [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+        this.objectMapper = objectMapper;
+        this.jobPeriodMs = jobPeriodMs;
+        this.zkPath = ZKPaths.makePath(NAKADI_JOBS_PATH, jobName);
+        this.log = LoggerFactory.getLogger("nakadi-job." + jobName);
+    }
+
+    public void runJobLocked(final Runnable action) {
+        log.info("Trying to run job...");
+
+        if (!timeToRunAction()) {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/job/ExclusiveJobWrapper.java]: Could you please elaborate on the approach to run job at certain time as well as checking that it is time to run the action? I am thinking of something like, run jobs at certain time and see if there is a lock, then act or not.

File: src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java [hunk 14]
```diff
@@ -2,26 +2,31 @@
...
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.List;
+import java.util.Optional;
+import java.util.UUID;
 
 @DB
 @Repository
 public class TimelineDbRepository extends AbstractDbRepository {
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: Shouldn't this change affect methods?
- listTimelinesOrdered
- listTimelinesOrdered(final String eventType)

File: src/main/java/org/zalando/nakadi/service/StorageService.java (hunk 8 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/StorageService.java]: I would love to do that but this PR already grew too much. But later we of course should remove this Result.

File: src/main/java/org/zalando/nakadi/service/StorageService.java (hunk 9 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/StorageService.java]: Before there used to be 2 actions inside. As now there was only one call to DB - no need in transaction any more, so I now removed this wrapper.

File: src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java (hunk 12 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java]: yes, I know. But the plan was that it will be implemented within the ticket for updating retention time :)
In this one https://github.com/zalando/nakadi/pull/621

File: src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java (hunk 10 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java]: Why?
When we are at this point the topic is already created.
So I thought that it will be a good place to add this when we update the previous timeline.

File: src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java (hunk 12 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java]: interesting, at least I was not aware about it, when implementing #621 :) then I will file another ticket to tie it together in different pr. wdyt?

Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java]: Ok, makes sense. 
We can do it as a separate PR when both of current PRs are merged.

File: src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java (hunk 10 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java]: It will be closer to the date, when topic was really created, so the clean up time will be more precise, honestly, so much time passed since I wrote it I do not even remember why :)

File: src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java (hunk 11 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java]: 1) If the topic deletion failed and timeline is marked as deleted - the topic will "hang in air"
2) If the topic is deleted and the timeline is failed to be marked as deleted - then it will try again the next time Job runs.

It seems that #1 can be improved but I don't really see how. In general we can change it so that if topic failed to be removed - we don't mark timeline as deleted - so that it will try the next time. But the problem can be that the topic failed to be removed because it doesn't exist (e.g. was removed by hand or whatever), then in this case the timeline will try to be cleaned each Job run (so it will get into this infinite loop).

Do you see any ways to improve that?

File: src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+            topicRepository.deleteTopic(timeline.getTopic());
+        } catch (final TopicDeletionException e) {
+            LOG.error("Failed to delete topic {} for expired timeline {}", timeline.getTopic(), timeline.getId(), e);
+        }
+    }
+
+    private void markTimelineDeleted(final Timeline timeline) {
+        try {
+            timeline.setDeleted(true);
+            timelineDbRepository.updateTimelime(timeline);
```
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java]: So here we have two failure cases:
1) Timeline failed to be updated in DB - the exception will be thrown and the cache will not try to be updated.
2) Timeline was updated in DB but the cache failed to be updated. Well there is not much to do in this case. Do you suggest to revert timeline update in this case.

File: src/main/java/org/zalando/nakadi/service/job/ExclusiveJobWrapper.java (hunk 13 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/job/ExclusiveJobWrapper.java]: It is done like that because the job can be executed on any of Nakadi instances and Nakadi instances have different start time so they have different timers for this Job.
So if I would implement it just to run each N minutes - then the job can be executed just few seconds after it was executed on another Nakadi instance - so the running of Job will be quite random. That's why I need this check if it's time to run job (it checks the timestamp when the Job was executed the last time)

File: src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java (hunk 10 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/timeline/TimelineService.java]: Such a precision will not make much sense here as the Job that cleans timelines is run once per hour.

File: src/acceptance-test/java/org/zalando/nakadi/webservice/TimelinesControllerAT.java (hunk 5 above)
Comment (v-stepanov) [inline src/acceptance-test/java/org/zalando/nakadi/webservice/TimelinesControllerAT.java]: @adyach In our acceptance tests we almost never clean (except user journey).

My opinion is that cleaning-after + tests-that-assume-the-DB-is-empty is a bad approach. It will always lead to some tests that forget to clean and some other tests fail because of that.

I think the better approach is not to clean after tests + all tests assume that DB is dirty + the tests that need clean-DB clean it in `@Before`. This is actually the approach we have now in Nakadi; I think for us it works fine.

File: src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java (hunk 11 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java]: My point of not understanding that is `topicRepository.deleteTopic` does not delete topic form Kafka, it just schedule it for deletion (maybe we should reconsider naming). Only exception it throws is UnknownTopicOrPartitionException (according to Kafka client code), so I think it is safe to not handle this. Methods renaming will be good.

File: src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java (hunk 15 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java]: Yes, if the update of the cache is failed, It is better to remove rollback timeline update.

File: src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java (hunk 14 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: This is true. It seems that inactive timelines are not filtered out currently. I will add that.

Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: I added filtering of inactive timelines.
https://github.com/zalando/nakadi/pull/611/commits/8fcce0bda1dba198b3bd93a01bacf58bbd320661

File: src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java (hunk 15 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java]: Ok, I added this functionality https://github.com/zalando/nakadi/pull/611/commits/69d1e9ed5a3589d65e93ff00e1a6253ba65a1adc

File: src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java [hunk 16]
```diff
@@ -65,16 +65,31 @@ private void deleteTimelineTopic(final Timeline timeline) {
     }
 
     private void markTimelineDeleted(final Timeline timeline) {
+        boolean timelineUpdatedInDB = false;
+        boolean cacheUpdated = false;
         try {
             timeline.setDeleted(true);
             timelineDbRepository.updateTimelime(timeline);
+            timelineUpdatedInDB = true;
```
Comment (adyach) [inline src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java]: Do you really need this? In my opinion only cacheUpdated is enough.

File: src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java (hunk 14 above)
Comment (adyach) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: Maybe we also should change listTimelinesOrdered, because it is used in the cache? In that case maybe it is better to do it on db level.

File: src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java (hunk 16 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/service/job/TimelineCleanupJob.java]: I need this to know if timeline was updated in DB. Because if it wasn't updated - then there is nothing to revert.

File: src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java (hunk 14 above)
Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: After I merge with new consuming code in master I will check again all use-cases. But for now it seems for me that it's better to do the filtering only on Service layer.

Comment (v-stepanov) [inline src/main/java/org/zalando/nakadi/repository/db/TimelineDbRepository.java]: I merged with latest master and checked all the use-cases, the only place where we need to filter out inactive timelines is `TimelineService.getActiveTimelinesOrdered`. In all other places it is ok to have all timelines (including expired ones).

**Comment by v-stepanov** [comment]
I think I now addressed all the comments.

**Comment by v-stepanov** [comment]
👍

**Comment by adyach** [comment]
👍

**Comment by adyach** [comment]
👍

**Comment by v-stepanov** [comment]
👍

**Comment by adyach** [comment]
👍
